core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java

```diff
@@ -176,6 +177,74 @@ public int getAddressIndex(int position, Page hashChannelsPage)
         return -1;
     }
 
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage, long[] rawHashes)
+    {
+        return getAddressIndex(positions, hashChannelsPage);
+    }
+
+    @Override
+    public int[] getAddressIndex(int[] positions, Page hashChannelsPage)
+    {
+        checkArgument(hashChannelsPage.getChannelCount() == 1, "Non-signle channel page passed to BigintPagesHash");
+
+        int positionCount = positions.length;
+        long[] incomingValues = new long[positionCount];
+        int[] hashPositions = new int[positionCount];
+
+        for (int i = 0; i < positionCount; i++) {
+            incomingValues[i] = hashChannelsPage.getBlock(0).getLong(positions[i], 0);
+            hashPositions[i] = getHashPosition(incomingValues[i], mask);
+        }
+
+        int[] found = new int[positionCount];
+        int foundCount = 0;
+        int[] result = new int[positionCount];
+        Arrays.fill(result, -1);
+        int[] foundKeys = new int[positionCount];
+
+        // Search for positions in the hash array. The ones that were found are put into `found` array,
+        // while the `foundKeys` arrays holds the keys that has been read from the hash array
+        for (int i = 0; i < positionCount; i++) {
+            if (key[hashPositions[i]] != -1) {
```

Comment:
This loop is one of the most expensive ops here due to random load from the (possibly large) `key` array.
There is a data dependency here between loop iterations due to `foundCount` depending on the `key` value.
I think splitting this into two loops, one that just calculates keys and the other that updates the `found` arrays, would be better here.

Reply:
That's an excellent idea. I will check it out.